program/src/logic.rs

```diff
@@ -382,6 +383,130 @@ pub fn distribute_fees<'a, 'b>(
     Ok(())
 }
 
+/// Checks if the stake accounts on `accounts` correspond to the ones generated
+/// by the validator's seeds. Returns the destination bump seed.
+pub fn check_unstake_accounts(
+    program_id: &Pubkey,
+    lido: &Lido,
+    accounts: &UnstakeAccountsInfo,
+) -> Result<u8, ProgramError> {
+    let validator = lido.validators.get(accounts.validator_vote_account.key)?;
+
+    // If a validator doesn't have a stake account, it cannot be unstaked.
+    if validator.entry.stake_seeds.stake_accounts_seed_begin
+        == validator.entry.stake_seeds.stake_accounts_seed_end
+    {
+        msg!("Attempting to unstake from accounts in a validator that has no stake accounts.");
+        return Err(LidoError::InvalidStakeAccount.into());
+    }
+    // Does not underflow. If `stake_accounts_seed_end == 0` the previous
+    // condition would have failed.
+    let source_stake_seed = validator.entry.stake_seeds.stake_accounts_seed_end - 1;
+    let destination_stake_seed = validator.entry.unstake_seeds.stake_accounts_seed_end;
+
+    let (source_stake_account, _) = Validator::find_stake_account_address(
+        program_id,
+        accounts.lido.key,
+        accounts.validator_vote_account.key,
+        source_stake_seed,
+    );
```

Comment:
Could we split this verification into calling the existing `Lido::check_stake_account` twice instead?

Reply:
If we do it like that, we'll have to pass the authority also for calculating stake/unstake addresses. I find it more complicated like that.

Reply:
Okay.